Source/NSFontCollection.m

```diff
@@ -0,0 +1,533 @@
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import <Foundation/NSArray.h>
+#import <Foundation/NSError.h>
+#import <Foundation/NSValue.h>
+#import <Foundation/NSDictionary.h>
+#import <Foundation/NSLock.h>
+#import <Foundation/NSKeyedArchiver.h>
+#import <Foundation/NSException.h>
+#import <Foundation/NSFileManager.h>
+#import <Foundation/NSPathUtilities.h>
+#import <Foundation/NSString.h>
+#import <AppKit/NSFontCollection.h>
+#import <GNUstepGUI/GSFontInfo.h>
+
+static NSMutableArray *_availableFontCollections = nil;
+static NSLock *_fontCollectionLock = nil;
+
+@interface NSFontCollection (Private)
+
++ (void) _loadAvailableFontCollections;
+- (BOOL) _writeToFile: (NSString *)path;
+- (void) _removeFile;
+- (void) _setFonts: (NSArray *)fonts;
+- (void) _setQueryAttributes: (NSArray *)queryAttributes;
+
+@end
+
+@implementation NSFontCollection (Private)
+
+/**
+ * Load all font collections....
+ */
++ (void) _loadAvailableFontCollections
+{
+  [_fontCollectionLock lock];
+  if (_availableFontCollections != nil)
+    {
+      // Nothing to do ... already loaded
+      [_fontCollectionLock unlock];
+    }
+  else
+    {
+      NSString			*dir;
+      NSString			*file;
+      NSEnumerator		*e;
+      NSFileManager		*fm = [NSFileManager defaultManager];
+      NSDirectoryEnumerator	*de;
+      NSFontCollection		*newCollection;
+
+      if (_availableFontCollections == nil)
+	{
+	  // Create the global array of font collections...
+	  _availableFontCollections = [[NSMutableArray alloc] init];
+	}
+      else
+	{
+	  [_availableFontCollections removeAllObjects];
+	}
+      
+      /*
+       * Load color lists found in standard paths into the array
+       * FIXME: Check exactly where in the directory tree we should scan.
+       */
+      e = [NSSearchPathForDirectoriesInDomains(NSLibraryDirectory,
+	NSAllDomainsMask, YES) objectEnumerator];
+
+      while ((dir = (NSString *)[e nextObject])) 
+	{
+	  BOOL flag;
+
+	  dir = [dir stringByAppendingPathComponent: @"FontCollections"];
+	  if (![fm fileExistsAtPath: dir isDirectory: &flag] || !flag)
+	    {
+	      // Only process existing directories
+	      continue;
+	    }
+
+	  de = [fm enumeratorAtPath: dir];
+	  while ((file = [de nextObject])) 
+	    {
+	      if ([[file pathExtension] isEqualToString: @"collection"])
+		{
+		  NSString	*name;
+		  name = [file stringByDeletingPathExtension];
+		  newCollection = [NSKeyedUnarchiver unarchiveObjectWithFile: [dir stringByAppendingPathComponent: file]];
+                  if (newCollection != nil)
+                    {
+                      [_availableFontCollections addObject: newCollection];
+                      RELEASE(newCollection);
```

Comment:
I am a bit unsure here but I would expect that objects returned from the keyed archiver are autoreleased. If this is true you are creating a zombie here.